### PR TITLE
feat(async): dispatcher

### PR DESCRIPTION
**Pull request by ocnc2**
## Summary by CodeRabbit

- **New Features**
  - Introduced a `MessageServer` for managing asynchronous messaging with structured error handling.
  - Added `Dispatcher` for improved asynchronous communication across services.
  - Implemented a `Publisher` type for event broadcasting.

- **Improvements**
  - Streamlined handling of finalized block events across multiple services.
  - Updated naming conventions for better clarity across packages.
  - Enhanced error management for subscription processes, improving system robustness.

- **Dependencies**
  - Updated module dependencies for enhanced functionality and performance.

### CODE REVIEW

File: mod/async/pkg/dispatcher/dispatcher.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+// It will error if the <event> type is inconsistent with the publisher
+// registered for the given eventID.
+func (d *Dispatcher) Publish(event types.MessageI) error {
+	return d.eventServer.Publish(event)
+}
+
+// Request dispatches the given request to the message server.
+// It will error if the <req> and <resp> types are inconsistent with the
+// route registered for the given messageID.
+func (d *Dispatcher) Request(req types.MessageI, resp any) error {
```
Comment (itsdevbear) [inline mod/async/pkg/dispatcher/dispatcher.go:70]: SendRequest

File: mod/async/pkg/dispatcher/dispatcher.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	return d.msgServer.Request(req, resp)
+}
+
+// Respond dispatches the given response to the message server.
+// It will error if the <resp> type is inconsistent with the route registered
+// for the given messageID.
+func (d *Dispatcher) Respond(resp types.MessageI) error {
```
Comment (itsdevbear) [inline mod/async/pkg/dispatcher/dispatcher.go:77]: SendResponse

File: mod/async/pkg/messaging/assert.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+// AN “AS IS” BASIS. LICENSOR HEREBY DISCLAIMS ALL WARRANTIES AND CONDITIONS,
+// EXPRESS OR IMPLIED, INCLUDING (WITHOUT LIMITATION) WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT, AND
+// TITLE.
+
+package messaging
+
+// ensureType ensures that the provided entity is of type T.
+// It returns a typed entity or an error if the type is not correct.
+func ensureType[T any](e any) (T, error) {
```
Comment (itsdevbear) [inline mod/async/pkg/messaging/assert.go:25]: can be done using generics

File: mod/async/pkg/dispatcher/types.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+package dispatcher
+
+import (
+	"context"
+
+	"github.com/berachain/beacon-kit/mod/async/pkg/types"
+	"github.com/berachain/beacon-kit/mod/log"
+)
+
+type MessageServer interface {
```
Comment (itsdevbear) [inline mod/async/pkg/dispatcher/types.go:30]: add some comments / can we make API surface smaller?

File: mod/beacon/blockchain/service.go [hunk 5]
```diff
@@ -165,50 +156,71 @@ func (s *Service[
...
+// VerifyBeaconBlock, and ProcessGenesisData requests, and handles them
+// accordingly.
 func (s *Service[
-	_, _, _, _, _, _, _, _, _, _, _,
+	_, BeaconBlockT, _, _, _, _, _, _, GenesisT, _, _,
 ]) Start(ctx context.Context) error {
-	subBlkCh, err := s.blkBroker.Subscribe()
-	if err != nil {
+	// register a channel as the receiver for FinalizeBeaconBlock:
+	finalizeBlkReqs := make(chan *asynctypes.Message[BeaconBlockT])
```
Comment (itsdevbear) [inline mod/beacon/blockchain/service.go:166]: put these on the service opposed to passing them around

File: mod/beacon/block_store/types.go [hunk 6]
```diff
@@ -42,7 +42,7 @@ type BlockStore[BeaconBlockT BeaconBlock] interface {
 // Event is an interface for block events.
 type Event[BeaconBlockT BeaconBlock] interface {
 	// Type returns the type of the event.
```
Comment (itsdevbear) [inline mod/beacon/block_store/types.go:44]: comment fix

File: mod/da/pkg/da/service.go [hunk 7]
```diff
@@ -128,25 +146,27 @@ func (s *Service[_, _, BlobSidecarsT, _, _]) handleBlobSidecarsProcessRequest(
...
-	msg *asynctypes.Event[BlobSidecarsT],
+// handleSidecarsVerifyRequest handles the SidecarsVerifyRequest event.
+// It verifies the sidecars and publishes a SidecarsVerified event.
+func (s *Service[_, _, BlobSidecarsT, _]) handleSidecarsVerifyRequest(
+	msg *asynctypes.Message[BlobSidecarsT],
 ) {
-	err := s.receiveSidecars(msg.Data())
+	var err error
+	// verify the sidecars.
+	err = s.verifySidecars(msg.Data())
```
Comment (archbear) [inline mod/da/pkg/da/service.go:169]: nit: put this in the if

File: mod/execution/pkg/deposit/service.go [hunk 8]
```diff
@@ -83,7 +83,7 @@ func NewService[
 		ExecutionPayloadT,
 		WithdrawalCredentialsT,
 	]{
-		feed:               feed,
+		finalizedBlockCh:   feed,
```
Comment (archbear) [inline mod/execution/pkg/deposit/service.go:86]: rename feed

File: mod/node-core/pkg/components/messages.go [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+		RouteFactory(messages.VerifyBeaconBlock),
+		RouteFactory(messages.FinalizeBeaconBlock),
+		RouteFactory(messages.ProcessGenesisData),
+		RouteFactory(messages.VerifySidecars),
+		RouteFactory(messages.ProcessSidecars),
+	}
+}
+
+// RouteFactory creates a new route for the given message ID.
+func RouteFactory(mID string) asynctypes.MessageRoute {
```
Comment (archbear) [inline mod/node-core/pkg/components/messages.go:62]: i think this factory is kinda superfluous and really like dense but also maybe i just cant read

File: mod/node-core/pkg/components/service_registry.go [hunk 10]
```diff
@@ -30,45 +30,35 @@ import (
...
+	ValidatorService  *ValidatorService
+	Dispatcher        *Dispatcher
 }
 
 // ProvideServiceRegistry is the depinject provider for the service registry.
 func ProvideServiceRegistry(
 	in ServiceRegistryInput,
 ) *service.Registry {
 	return service.NewRegistry(
+		in.Dispatcher,
```
Comment (archbear) [inline mod/node-core/pkg/components/service_registry.go:52]: should the service registry force the requirement of having a dispatcher? in theory (very loose in theory) someone using as an sdk doesn't need to adhere to any of our async service logic

**Review by archbear — COMMENTED** [review]
mostly nits

File: mod/async/pkg/messaging/publisher.go [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+				}
+			}
+			return
+		case msg := <-p.msgs:
+			// broadcast published msg to all clients
+			for client := range p.clients {
+				// send msg to client (or discard msg after timeout)
+				select {
+				case client <- msg:
+				case <-time.After(p.timeout):
```
[the remark this answers is not in the document]
Comment (itsdevbear) [reply mod/async/pkg/messaging/publisher.go:83]: agree

File: mod/async/pkg/server/event.go [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+)
+
+// EventServer asyncronously dispatches events to subscribers.
+type EventServer struct {
+	publishers map[types.MessageID]types.Publisher
+	logger     log.Logger[any]
+}
+
+// NewEventServer creates a new event server.
+func NewEventServer() *EventServer {
```
[the remark this answers is not in the document]
Comment (ocnc2) [reply mod/async/pkg/server/event.go:37]: we could, but i want all dispatcher related logging to share a single logger instance with service="dispatcher"

**Review by archbear — COMMENTED** [review]
nothing to comment just stat padding

**Comment by ocnc2** [comment]
closing in favour of https://github.com/berachain/beacon-kit/pull/1899

**Pull request closed without merging**